### ISSUE TITLE
HTML containers and headers are not consistently structured or labeled

### ISSUE THREAD

**Issue by ssciolla** [tl-its-umich-edu/my-learning-analytics#807]
## Expected behavior

The structuring and labeling of container and heading HTML element should be consistent across all pages of the application.

## Describe the issue

From the 10/28/19 accessibility review:

<details>
<summary>(See review documents…)</summary>

> * MyLA Accessibility Review 10/28/18
>   * Google original: https://docs.google.com/document/d/1IFBNWHvzY1l8YT7G1TX9jCrOyU9yAxF1-KazNbb4kc0
>   * GitHub PDF copy: https://github.com/tl-its-umich-edu/my-learning-analytics/files/5502938/MyLA.Accessibility.Review.10_28_18.pdf
> * MyLA Accessibility Table 10/28/18 - Issue Notes
>   * Google original: https://docs.google.com/spreadsheets/d/1ZWB6Ip4I9H8718G4uMn6VcRDrpS7dkS5BCR7nB0Q8ao
>   * GitHub PDF copy: https://github.com/tl-its-umich-edu/my-learning-analytics/files/5502940/MyLA.Accessibility.Table.10_28_18.-.Issue.Notes.pdf

</details>

"Content is indifferently structured. This becomes important for the blind persona. … Header and footer are identified programmatically. The main content is not, however. … In addition - heading structure is a bit off. It starts with an h6 (we assume because it is embedded in Canvas) - but it has h5 subsections - at least the Progress toward Final Grade view does. This needs to be normalized."

## Additional context

~~Regarding the main container, the review suggests adding the `role="main"` attribute-value pair to the HTML element serving as the main container. The header problem might be solved simultaneously to issue #801 about the view headers.~~

WRT roles, new issue #1165 has been opened for that.

**Comment by lsloan** [comment]
@ssciolla, am I correct in understanding that the main points of this issue are:

* h5 sections as children of h6 ones
* add `role="main"` to the main HTML element

**Comment by ssciolla** [comment]
@lsloan, for the second bullet point, I think so? May want to read the W3C notes about that `role` type and how it should be used.

For the first bullet, I believe I at least partially resolved this in the PR linked to above (#949). I think the suggestion presented by Gonzalo was that we shouldn't be skipping header levels, i.e. that we should start with `h1` and then descend. I set up a level mapping in `globals.js` in that PR so that `h5` actually appears in the DOM as `h1`, and `h6` as `h2`. This should be in place for the visualization views, but we probably should look at how content is structured on the various menu and other pages. It's probably a good idea to have some kind of header on all pages, again to help give those using screenreaders a better sense of where they are in the application.

https://github.com/tl-its-umich-edu/my-learning-analytics/blob/0b1bb31fc68e3519a442f2b3afdd2a9dd83c195e/assets/src/globals.js#L32-L41

**Comment by ssciolla** [comment]
Accessibility checkers may be useful for figuring out some of these issues. I've heard of people on the UX team using AXE, though I haven't tried myself. I think this is what they were referring to: https://www.deque.com/axe/

**Comment by lsloan** [comment]
@gsilver, can you add some clarification to this issue?

@ssciolla made some changes that may resolve the header level problem.  I plan to talk with him about that in more detail soon.

Can you explain more about the need for the `role="main"` attribute?  What can I do to verify the problem exists and then verify it is corrected after I've attempted to fix it?  Should I use axe (appears to be a commercial product with a cost) or Chrome accessibility tools?

**Comment by lsloan** [comment]
This morning, I met with @gsilver and @ssciolla about this issue.

Prior to the meeting, Gonzalo emailed…

> "in the dashboard the heading structure is offset. I would make the "About ...." components a simple link without an h4 parent.

<details>
<summary>He included the following screenshot to illustrate the problem…</summary>

![image](https://user-images.githubusercontent.com/20786/102101652-e51d0480-3df8-11eb-8ecc-e273ce968417.png)

</details>

Sam suggested that the `h4` elements immediately within `h2` ones could be resolved in the code at (https://github.com/tl-its-umich-edu/my-learning-analytics/blob/4cabdbf34e3743894dba859a6e8b3e092c55e17d/assets/src/components/SelectCard.js#L157-L163).

Perhaps removing `component='h4'` would do the job.  If that leaves the element as `h5`, then it would require investigation to remove the headers and use some other styling instead.

We agreed that the roles part of the issue should be a separate issue, so I will open a new one for that.

**Comment by jennlove-um** [comment]
@lsloan I'm not sure how to test this one.  Can you give some guidance?

**Comment by ssciolla** [comment]
@jennlove-um (@lsloan), this issue has a lengthy history, but for the latest change, you can do this to test.

- Right-click on an information icon ("i") link in the list of visualizations, and go to "Inspect" or comparable utility to view the page's HTML.
- Make sure that the element doesn't use an `h4` tag and uses a `p` tag instead.

Please let me know if you would like additional guidance.

**Comment by jennlove-um** [comment]
@ssciolla Thanks for the clear steps.

Testing passes in myla-test - the p tags are there instead of the h4 tags that show up in production currently.
